ADSCompStat pipeline release candidate

seasidesparrow opened the pull request:
- Added documentation (`README.md`)
- Added unittests
- Improved matching logic
- Substantially reduced memory usage
- Improvements to tasks, Celery tasking
- Job batching for improved efficiency

seasidesparrow committed:
modifies format of completeness export
 	modified:   adscompstat/tasks.py

seasidesparrow committed:
modified:   alembic/versions/d2c43086a8ab_add_matching_data_tables.py
 	modified:   pyproject.toml

seasidesparrow committed:
fix: prevent 8-digit issnString from being interpreted as int()
 	modified:   adscompstat/tasks.py

seasidesparrow committed:
adds task to reprocess mismatches
 	modified:   adscompstat/tasks.py

seasidesparrow committed:
adds task to reprocess unmatched records
 	modified:   adscompstat/tasks.py

seasidesparrow committed:
added retries to run.py, fixed bug in sqlalchemy queries
 	modified:   adscompstat/tasks.py
 	modified:   run.py

seasidesparrow committed:
added bibstem synonyms to config.py
 	modified:   config.py

seasidesparrow committed:
added match logic to look for related bibstems
 	modified:   adscompstat/match.py
 	modified:   config.py
 	modified:   run.py

seasidesparrow committed:
added another journal to related bibstems
 	modified:   config.py

seasidesparrow committed:
started adding unittests, adjusted match.py logic

seasidesparrow committed:
updated README.md, minor change to logging in run.py
 	modified:   README.md
 	modified:   run.py

seasidesparrow committed:
removing .tmp
 	modified:   .gitignore
 	deleted:    .tmp/.coverage
 	deleted:    .tmp/v/cache/lastfailed
 	deleted:    .tmp/v/cache/nodeids
 	deleted:    .tmp/v/cache/stepwise

seasidesparrow committed:
Updating readme
 	modified:   README.md

seasidesparrow committed:
added more unit tests, minor edits to code
 	modified:   adscompstat/match.py
 	modified:   adscompstat/utils.py
 	modified:   run.py
 	new file:   tests/stubdata/input/all_list
 	new file:   tests/stubdata/input/alternate_list
 	new file:   tests/stubdata/input/canonical_list
 	new file:   tests/stubdata/input/deleted_list
 	new file:   tests/stubdata/input/doi_links
 	new file:   tests/stubdata/input/issn_bibstems
 	modified:   tests/test_match.py
 	modified:   tests/test_utils.py

ehenneken commented on adscompstat/match.py:
```diff
@@ -14,8 +14,22 @@
 class CrossrefMatcher(object):
-    def __init__(self):
-        pass
+    def __init__(self, related_bibstems=conf.get("RELATED_BIBSTEMS", [[]])):
```
Assigning a configuration value as argument in __init__ is not a pattern we typically use. Let's follow the pattern of assigning the default "empty" value and get the "conf" value in the init function itself

seasidesparrow replied:
I opted to do this differently.  Previously it was getting this data from config.py, and we do not normally store curated data in the pipeline or configurations.  Now, I'm loading this variable a single time in `tasks.py` from the json file `abstracts/config/journalsdb/PIPELINE/data/related_bibstems.json`.  Each time the CrossrefMatcher is instantiated in tasks, the variable is passed as an argument.  Loading it in tasks means the json will only need to be loaded once and then the tasks will have it in memory.

ehenneken commented on adscompstat/match.py:
```diff
@@ -14,8 +14,22 @@
+    def __init__(self, related_bibstems=conf.get("RELATED_BIBSTEMS", [[]])):
+        self.related_bibstems = related_bibstems
+
+    def _compare_bibstems(self, testBibstem, classicBibstem):
```
I haven't seen the rest of the code yet, so this may not be a relevant question: is it guaranteed here that bibstems are always compared in the same "form"? With "form" I mean that bibstems sometimes are padded with periods and sometimes they are not (e.g. ApJ vs ApJ..)

seasidesparrow replied:
The comparisons are done using the substring `bibcode[4:9]`, so these will include the trailing dots for bibstems less than five characters long.  Ultimately I would like to serve these data from the journals database rather than creating the `related_bibstems.json` file by hand, so I will keep this in mind when I automate this.

ehenneken commented on adscompstat/tasks.py:
```diff
@@ -453,3 +459,49 @@ def task_export_completeness_to_json():
+                logger.debug("Calling task_process_meta with batch '%s'" % batch)
+                task_process_meta.delay(batch)
+        except Exception as err:
+            logger.warning("Error reprocessing unmatched records: %s" % err)
```
It looks like the functions `task_retry_mismatched` and `task_retry_unmatched` can be made into one function where you specify the "matchtype" as argument? Is there reason why it's better to keep two separate functions?

seasidesparrow replied:
This is a great idea.  I changed this to accept a string containing the record's `matchtype`, and changed `run.py` so that calling this option will iteratively call this task with the strings `mismatch`, `unmatched`, and `failed`.  `Failed` was not in the original PR but should be, if (for example) a bugfix is made to `adsingestp.parsers.crossref` that will address reasons why the records failed.

ehenneken reviewed (changes requested):
Looks good. I left 3 comments for you to consider.

seasidesparrow committed:
changes requested in PR review
 	modified:   adscompstat/match.py
 	modified:   adscompstat/tasks.py
 	modified:   config.py
 	modified:   run.py

seasidesparrow committed:
Expanded readme
  Changes to be committed:
 	modified:   README.md

seasidesparrow commented:
I finished the README.md, including suggested information on how to run the code, along with suggested cronjobs for maintenance.

seasidesparrow committed:
updates ADSIngest repos to most recent releases, ran pre-commit
 	modified:   adscompstat/tasks.py
 	modified:   adscompstat/utils.py
 	modified:   config.py
 	modified:   pyproject.toml
 	modified:   run.py
 	modified:   tests/stubdata/input/canonical_list
 	modified:   tests/test_utils.py

kelockhart commented on alembic/versions/d2c43086a8ab_add_matching_data_tables.py:
I didn't have a good place to put it, but check line 90 in the `models.py` - looks like that line didn't get fully updated when you renamed a column.

seasidesparrow replied:
Fixed.

kelockhart commented on run.py:
Small suggestion - there are a few spots in here with some lines of code inside a `try` that won't fail (like line 93, `newlogs = list()`) - consider editing so that only the lines that might fail are inside the `try`. It'll make debugging easier, if an exception gets thrown.

kelockhart commented on adscompstat/tasks.py:
```diff
@@ -31,6 +32,18 @@
+    logger.warning("Related bibstems filename not set.")
+
 
 # No delay/queue, synchronous only
```
Generally, the tasks in `tasks.py` are all async celery tasks - the commenting is clear and makes it obvious that these aren't running via rabbitmq, but I'm curious why these synchronous functions are here instead of at the top of `run.py` or something. Is there a possibility these get moved to an async process at some point?

seasidesparrow replied:
Two reasons: one, that I didn't want to instantiate a sql session manager in both run.py and tasks.py; and two, I was playing around with trying to load the data asynchronously, especially the lists of bibcodes.  I never got it to work correctly here, but eventually that data store going to be in a separate identifiers database rather than in compstat.  When we have that in place, I'll remove these specific tasks.

kelockhart commented on adscompstat/tasks.py:
Why is `task_write_matched_record_to_db` an async task? Seems like it'd be pretty fast - is it slower than I think?

seasidesparrow replied:
Under very heavy load (full reindexing with a large number of matching workers) I did see some messages collect in this queue, but typically not under weekly-update level loads.  I could remove this if you prefer.

kelockhart commented on adscompstat/tasks.py:
```diff
@@ -453,3 +471,25 @@ def task_export_completeness_to_json():
             logger.error("Unable to export completeness data to disk: %s" % err)
+
+
+@app.task(queue="get-logfiles")
```
Did you mean to have this task as async? It seems fast, spinning off async tasks for the heavier parts, and this queue is already being used for another task.

seasidesparrow replied:
Keeping it async will let it process the unmatched, mismatched, and failed records simultaneously, in the same way that the other task using this queue will process each logfile simultaneously.  The two tasks will never run at the same time so I thought reusing the queue would be more efficient, but I can make a new queue specifically for this for clarity.

kelockhart commented on tests/test_match.py:
No where else to put this, so I'm sticking this comment here - will there by testing on `tasks.py`? There's a good amount of logic in there.

seasidesparrow replied:
Yes, I ran out of time, and don't have experience mocking celery and sqlalchemy calls yet.  I'd like to get this into production now and circle back.  It may become moot anyway if we move from celery-based tasks to gRPC(?).

kelockhart reviewed (commented):
Looks good as far as efficiency, and all the previous concerns were addressed nicely. My comments and suggestions are more on the level of code style, but it's good to go from a high-level perspective.

seasidesparrow committed:
Fix: issue #14 (completeness summary exception)
 	modified:   adscompstat/models.py
 	modified:   adscompstat/tasks.py
 	modified:   run.py
 	new file:   tests/test_tasks.py

seasidesparrow committed:
modified:   adscompstat/tasks.py
 	modified:   config.py
 	new file:   tests/stubdata/input/10.1186:300627.long-fake.2023-12-10
 	new file:   tests/stubdata/input/10.1186:300627.out.2023-12-10
 	modified:   tests/test_tasks.py
